Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs

```diff
@@ -1,32 +1,152 @@
         {
-            if (!comp.AlreadyEmagged)
+            if (!reagentDispenser.IsEmagged)
             {
-                comp.AddFromPrototype(comp.EmagPackPrototypeId);
-                comp.AlreadyEmagged = true;
+                reagentDispenser.IsEmagged = true;
                 args.Handled = true;
+                UpdateUiState(reagentDispenser);
+            }
+        }
+
+        private void OnSetDispenseAmountMessage(EntityUid uid, ReagentDispenserComponent reagentDispenser, ReagentDispenserSetDispenseAmountMessage message)
+        {
+            reagentDispenser.DispenseAmount = message.ReagentDispenserDispenseAmount;
+            UpdateUiState(reagentDispenser);
+            ClickSound(reagentDispenser);
+        }
+
+        private void OnDispenseReagentMessage(EntityUid uid, ReagentDispenserComponent reagentDispenser, ReagentDispenserDispenseReagentMessage message)
+        {
+            // Ensure that the reagent is something this reagent dispenser can dispense.
+            if (!GetInventory(reagentDispenser).Contains(message.ReagentId))
+                return;
+
+            var outputContainer = _itemSlotsSystem.GetItem(reagentDispenser.Owner, SharedReagentDispenser.OutputSlotName);
+            if (outputContainer is not {Valid: true} || !_solutionContainerSystem.TryGetFitsInDispenser(outputContainer.Value, out var solution))
+                return;
+
+            if (_solutionContainerSystem.TryAddReagent(outputContainer.Value, solution, message.ReagentId, (int)reagentDispenser.DispenseAmount, out var dispensedAmount)
+                && message.Session.AttachedEntity is not null)
+            {
+                _adminLogger.Add(LogType.ChemicalReaction, LogImpact.Medium,
+                    $"{_entities.ToPrettyString(message.Session.AttachedEntity.Value):player} dispensed {dispensedAmount}u of {message.ReagentId} into {_entities.ToPrettyString(outputContainer.Value):entity}");
             }
+
+            UpdateUiState(reagentDispenser);
+            ClickSound(reagentDispenser);
+        }
+
+        private void OnClearContainerSolutionMessage(EntityUid uid, ReagentDispenserComponent reagentDispenser, ReagentDispenserClearContainerSolutionMessage message)
+        {
+            var outputContainer = _itemSlotsSystem.GetItem(reagentDispenser.Owner, SharedReagentDispenser.OutputSlotName);
+            if (outputContainer is not {Valid: true} || !_solutionContainerSystem.TryGetFitsInDispenser(outputContainer.Value, out var solution))
+                return;
+
+            _solutionContainerSystem.RemoveAllSolution(outputContainer.Value, solution);
+            UpdateUiState(reagentDispenser);
+            ClickSound(reagentDispenser);
+        }
+
+        private void ClickSound(ReagentDispenserComponent reagentDispenser)
+        {
+            _audioSystem.Play(reagentDispenser.ClickSound, Filter.Pvs(reagentDispenser.Owner), reagentDispenser.Owner, AudioParams.Default.WithVolume(-2f));
```

Comment:
```suggestion
            _audioSystem.PlayPvs(reagentDispenser.ClickSound, reagentDispenser.Owner, AudioParams.Default.WithVolume(-2f));
```
might not be the exact api idr